File: HW/HW9/Hw9.py
class TreeNode(object):

  def __init__(self, val=0, left=None, right=None):

    self.val = val

    self.left = left

    self.right = right


def lowestCommonAncestor(root: 'TreeNode | None', p: 'TreeNode',
                         q: 'TreeNode'):
  if root == None:
    return None
  if root == p or root == q:
    return root
  left = lowestCommonAncestor(root.left, p, q)
  right = lowestCommonAncestor(root.right, p, q)
  if left and right:
    return root
  return left if left else right


# Create leaf nodes
node1 = TreeNode(1)
node5 = TreeNode(5)

# Create node 2 and attach 1 as a child
node2 = TreeNode(2, node1, None)

# Create node 4 and attatch 5 as child
node4 = TreeNode(4, None, node5)

# Create the root node and attach 2 and 4 as children
root = TreeNode(3, node2, node4)

printNode1 = lowestCommonAncestor(root, node1, node5)
printNode2 = lowestCommonAncestor(root, node1, node2)
printNode3 = lowestCommonAncestor(root, node2, node4)
printNode4 = lowestCommonAncestor(root, root, node5)

if (printNode1 != None):
  print(printNode1.val)
if (printNode2 != None):
  print(printNode2.val)
if (printNode3 != None):
  print(printNode3.val)
if (printNode4 != None):
  print(printNode4.val)

print()
print("Testing the tree provided in the assignment")
print()

# Create leaf nodes
nodes1 = TreeNode(1)
nodes2 = TreeNode(2)
nodes5 = TreeNode(5)
nodes9 = TreeNode(9)

# Create nodes 2nd row
nodes3 = TreeNode(3, nodes1, nodes2)
nodes8 = TreeNode(8, nodes5, nodes9)

# Create node 4 and attatch 5 as child
roots4 = TreeNode(4, nodes3, nodes8)

printNodes1 = lowestCommonAncestor(roots4, nodes3, nodes1)
printNodes2 = lowestCommonAncestor(roots4, nodes1, nodes5)
printNodes3 = lowestCommonAncestor(roots4, nodes2, nodes3)
printNodes4 = lowestCommonAncestor(roots4, nodes8, nodes5)

if (printNodes1 != None):
  print(printNodes1.val)
if (printNodes2 != None):
  print(printNodes2.val)
if (printNodes3 != None):
  print(printNodes3.val)
if (printNodes4 != None):
  print(printNodes4.val)

File: HW/HW9/Hw9.java
class Hw9 {
	public static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
		if(root == null) {
			return null;
		}
		if (p == root || q == root) {
			return root;
		}
		TreeNode left = lowestCommonAncestor(root.left, p, q);
		TreeNode right = lowestCommonAncestor(root.right, p, q);
		if(left != null && right != null) {
			return root;
		}
		else if (left != null) {
			return left;
		} else {
			return right;
		}
	}
	
	public static void main(String[] args) {
		// Create leaf nodes
		TreeNode nodes1 = new TreeNode(1);
		TreeNode nodes2 = new TreeNode(2);
		TreeNode nodes5 = new TreeNode(5);
		TreeNode nodes9 = new TreeNode(9);

		// Create nodes 2nd row
		TreeNode nodes3 = new TreeNode(3, nodes1, nodes2);
		TreeNode nodes8 = new TreeNode(8, nodes5, nodes9);

		// Create node 4 and attach 5 as child
		TreeNode roots4 = new TreeNode(4, nodes3, nodes8);

		TreeNode printNodes1 = lowestCommonAncestor(roots4, nodes3, nodes1);
		TreeNode printNodes2 = lowestCommonAncestor(roots4, nodes1, nodes5);
		TreeNode printNodes3 = lowestCommonAncestor(roots4, nodes2, nodes3);
		TreeNode printNodes4 = lowestCommonAncestor(roots4, nodes8, nodes5);

		System.out.println(printNodes1.val);
		System.out.println(printNodes2.val);
		System.out.println(printNodes3.val);
		System.out.println(printNodes4.val);
	}
}


class TreeNode {

	int val;

	TreeNode left;

	TreeNode right;

	TreeNode() {}

	TreeNode(int val) { this.val = val; }

	TreeNode(int val, TreeNode left, TreeNode right) {

		this.val = val;

		this.left = left;

		this.right = right;

	}
}
